ssl/t1_lib.c

```diff
@@ -866,6 +872,21 @@ static const SIGALG_LOOKUP *tls1_get_legacy_sigalg(const SSL *s, int idx)
     if (SSL_USE_SIGALGS(s) || idx != SSL_PKEY_RSA) {
         const SIGALG_LOOKUP *lu = tls1_lookup_sigalg(tls_default_sigalg[idx]);
 
+        /*
+         * Some GOST ciphersuites allow more than one signature algorithms
+         * */
+        if ((start_idx == -1) && (s->server) && (idx == SSL_PKEY_GOST01)
+            && (s->s3->tmp.new_cipher->algorithm_auth != SSL_aGOST01)) {
+            int real_idx;
+            for (real_idx = SSL_PKEY_GOST01; real_idx <= SSL_PKEY_GOST12_512;
+                 real_idx++) {
+                if (s->cert->pkeys[real_idx].privatekey != NULL) {
+                    break;
+                }
+            }
+            lu = tls1_lookup_sigalg(tls_default_sigalg[real_idx]);
```

Comment:
I'm wondering if this whole block might be better placed in the section labelled "/* Work out index corresponding to ciphersuite */" above? It seems that we work out a value for idx there, and then change our mind here. That would also mean we don't call "tls1_lookup_sigalg" twice.